This notebook is the next step in cleaning the data + feature engineering. Unlike the python script that conduct dated cleaning for "obvious" things like fixing data types, (E.g. replacing 19.2% with 0.192) and added measures like calculating lost principle on defaulted loans and monthly debt payments per customer, this one conducts a more in depth study to figure out how to handle more complicated missing values, condensing certain categorical variables and calculating a high level default rate.

**High Level Findings**
* Default Rate 19.2% 
* Average Loan 15,432.11
* Average Lost principle on a defaulted loan: 2,048.37
* Average interest rate 12.6%

Beyond the above, I won't dig much more into the details of LC's loan in order to respect their rules around how to use thier data. 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt  
import seaborn as sn
%matplotlib inline

In [2]:
# set parameters for how many viewable columns there are 

pd.set_option('max_columns', None)
pd.set_option('display.max_rows', None)


In [3]:
lending = pd.read_csv('data/2015_processed_log(4).csv')
lending.shape

(378516, 66)

In [4]:
lending.head()

funded_amnt        term  int_rate  installment grade  emp_length  \
0      20000.0   36 months    0.1485       691.84     C           6   
1      20000.0   36 months    0.1577       700.88     D           5   
2      10000.0   60 months    0.1797       253.78     D           2   
3      20000.0   36 months    0.0849       631.26     B          10   
4      10000.0   36 months    0.0649       306.45     A           6   

  home_ownership  annual_inc verification_status     issue_d  loan_status  \
0           RENT    110000.0        Not Verified  2015-12-01   Fully Paid   
1           RENT     70000.0        Not Verified  2015-12-01   Fully Paid   
2       MORTGAGE     55000.0        Not Verified  2015-12-01  Charged Off   
3       MORTGAGE     85000.0        Not Verified  2015-12-01   Fully Paid   
4           RENT     85000.0        Not Verified  2015-12-01   Fully Paid   

          purpose addr_state    dti  delinq_2yrs earliest_cr_line  \
0     credit_card         IL  12.45          0.0       2007-06-01   
1           house         FL  22.21          0.0       2004-07-01   
2     credit_card         CO  35.70          0.0       2001-04-01   
3  major_purchase         SC  17.61          1.0       1999-02-01   
4     credit_card         PA  13.07          0.0       2002-04-01   

   fico_range_low  inq_last_6mths  mths_since_last_delinq  open_acc  pub_rec  \
0           690.0             0.0                     NaN       8.0      0.0   
1           680.0             0.0                    38.0      24.0      0.0   
2           685.0             0.0                     NaN      14.0      0.0   
3           705.0             0.0                     3.0       8.0      0.0   
4           685.0             1.0                     NaN      14.0      1.0   

   pub_rec_bankruptcies  revol_bal  revol_util  total_acc  total_pymnt  \
0                   0.0    21374.0       0.845       12.0  24889.01336   
1                   0.0    19077.0       0.366       63.0  21780.58678   
2                   0.0    38623.0       0.780       28.0   5558.20000   
3                   0.0      826.0       0.057       15.0  21538.50898   
4                   1.0    10464.0       0.345       23.0  10998.97157   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         24889.01         20000.00        4889.01                 0.0   
1         21780.59         20000.00        1780.59                 0.0   
2          5558.20          2687.15        2871.05                 0.0   
3         21538.51         20000.00        1538.51                 0.0   
4         10998.97         10000.00         998.97                 0.0   

   collections_12_mths_ex_med application_type  acc_now_delinq  \
0                         0.0       Individual             0.0   
1                         0.0       Individual             0.0   
2                         0.0       Individual             0.0   
3                         0.0       Individual             0.0   
4                         0.0       Individual             0.0   

   chargeoff_within_12_mths  acc_open_past_24mths  avg_cur_bal  delinq_amnt  \
0                       0.0                   3.0       5356.0          0.0   
1                       0.0                   9.0       2759.0          0.0   
2                       0.0                   4.0      20578.0          0.0   
3                       0.0                   4.0      17700.0          0.0   
4                       0.0                   7.0       1997.0          0.0   

   mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  \
0                 102.0                   16.0            10.0       0.0   
1                 137.0                    2.0             2.0       1.0   
2                 176.0                    7.0             6.0       6.0   
3                  55.0                   32.0            13.0       3.0   
4                 129.0                    1.0             1.0       1.0   

Get high level summary stats before we proceed further with cleaning and EDA

In [5]:
lending.describe()

funded_amnt       int_rate    installment     emp_length  \
count  378516.000000  378516.000000  378516.000000  378516.000000   
mean    15436.261809       0.125736     446.415025       6.114080   
std      8590.628539       0.043238     245.451198       3.577084   
min      1000.000000       0.053200      14.010000       1.000000   
25%      9000.000000       0.091700     267.647500       3.000000   
50%     14050.000000       0.122900     390.730000       6.000000   
75%     20250.000000       0.155900     586.000000      10.000000   
max     35000.000000       0.289900    1445.460000      10.000000   

         annual_inc            dti    delinq_2yrs  fico_range_low  \
count  3.785160e+05  378516.000000  378516.000000   378516.000000   
mean   7.856514e+04      19.070799       0.351626      693.687968   
std    7.616610e+04       8.653231       0.933041       30.009819   
min    3.800000e+03       0.000000       0.000000      660.000000   
25%    4.800000e+04      12.550000       0.000000      670.000000   
50%    6.600000e+04      18.510000       0.000000      685.000000   
75%    9.400000e+04      25.210000       0.000000      710.000000   
max    9.500000e+06     380.530000      39.000000      845.000000   

       inq_last_6mths  mths_since_last_delinq       open_acc        pub_rec  \
count   378516.000000           195602.000000  378516.000000  378516.000000   
mean         0.570005               33.846765      12.029959       0.227465   
std          0.866236               21.970307       5.651834       0.644780   
min          0.000000                0.000000       1.000000       0.000000   
25%          0.000000               15.000000       8.000000       0.000000   
50%          0.000000               30.000000      11.000000       0.000000   
75%          1.000000               50.000000      15.000000       0.000000   
max          6.000000              176.000000      82.000000      86.000000   

       pub_rec_bankruptcies     revol_bal     revol_util      total_acc  \
count         378516.000000  3.785160e+05  378516.000000  378516.000000   
mean               0.128753  1.785511e+04       0.539537      25.541903   
std                0.380468  2.437875e+04       0.238838      12.095716   
min                0.000000  0.000000e+00       0.000000       4.000000   
25%                0.000000  6.572000e+03       0.362000      17.000000   
50%                0.000000  1.213500e+04       0.543000      24.000000   
75%                0.000000  2.164000e+04       0.723000      32.000000   
max                9.000000  2.904836e+06       1.930000     169.000000   

         total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
count  378516.000000    378516.000000    378516.000000  378516.000000   
mean    16790.698706     16783.933857     13385.621481    3111.864650   
std     10868.747671     10864.168847      8861.575622    3159.109501   
min         0.000000         0.000000         0.000000       0.000000   
25%      8304.155000      8301.377500      6000.000000    1006.030000   
50%     14360.042135     14352.540000     12000.000000    2001.815000   
75%     23131.887758     23123.845000     19750.000000    4124.750000   
max     65659.902030     65659.900000     35000.000000   30659.900000   

       total_rec_late_fee  collections_12_mths_ex_med  acc_now_delinq  \
count        3.785160e+05               378516.000000   378516.000000   
mean         2.127155e+00                    0.020496        0.006103   
std          1.495169e+01                    0.158915        0.086710   
min         -5.100000e-09                    0.000000        0.000000   
25%          0.000000e+00                    0.000000        0.000000   
50%          0.000000e+00                    0.000000        0.000000   
75%          0.000000e+00                    0.000000        0.000000   
max          1.115370e+03                   14.000000       14.000000   

       chargeoff_within_12_mths  acc_open_past_24mths    avg_

Analyze the loan purpose column, as there may be opportunities to condense the categories. 

In [6]:
# count the values in each column 

purpose_count = pd.DataFrame(lending['purpose'].value_counts())

# resetting the index because the values in the purpose column will get set as the index

purpose_df = purpose_count.reset_index()

# fix the true index vs. categories columns 

purpose_df.rename(columns={'purpose':'count'}, inplace = True)
purpose_df.rename(columns={'index':'loan_purpose'}, inplace = True)

# next add % of total 

loan_count = len(lending)

purpose_df['per_of_total'] = purpose_df['count'] / loan_count

purpose_df

loan_purpose   count  per_of_total
0   debt_consolidation  225001      0.594429
1          credit_card   91789      0.242497
2     home_improvement   22600      0.059707
3                other   16994      0.044896
4       major_purchase    6743      0.017814
5              medical    3452      0.009120
6                  car    3145      0.008309
7       small_business    3092      0.008169
8               moving    2175      0.005746
9             vacation    1998      0.005279
10               house    1323      0.003495
11    renewable_energy     199      0.000526
12             wedding       4      0.000011
13         educational       1      0.000003

* Caveat: all of these purposes are self_reported, so they may not have much value. People could be reporting the category that they think makes it most likely that they'll get the loan, rather than their true purpose. 
* ~83.8% of the customers were in the debt consolidation or credit card category, which are effectively the same as debt consolidation and taking a loan to pay off credit cards is effectively the same thing. 
* ~90% of customers were taking out loans to consolidate debt or for home improvements. 
* The other categories are too small to be relevant from a modeling perspective, will combine the categories as follows, condensing 14 categories into three. 
    * debt consolidation & credit card into 'debt_consolidation'
    * consumer_credit made up of home_improvement, major_purchase, car, house, vacation, moving 
    * rest will go into "other" 


In [7]:
# the code to condense categories will hot encode them at the same time 

lending.loc[:, 'debt_consolidation'] = lending.loc[:,'purpose'].apply(lambda x:\
            1 if x == 'debt_consolidation' else (1 if x == 'credit_card' else 0))

lending.loc[:, 'consumer_credit'] = lending.loc[:,'purpose'].apply(lambda x: 1 if\
            x == 'home_improvement' else (1 if x == 'major_purchase' else 1 if x\
                                          == 'car' else 1 if x == 'wedding' else 1 if x == 'moving' else 1 if x\
                                          == 'house' else 1 if x == 'vacation' else  0))

lending.loc[:, 'other'] = lending.loc[:,'purpose'].apply(lambda x: 1 if x\
            == 'other' else (1 if x == 'medical' else 1 if x == 'small_business' else\
                             1 if x == 'renewable_energy' else 1 if x == 'educational' else 0))

In [8]:
lending.head(5)

funded_amnt        term  int_rate  installment grade  emp_length  \
0      20000.0   36 months    0.1485       691.84     C           6   
1      20000.0   36 months    0.1577       700.88     D           5   
2      10000.0   60 months    0.1797       253.78     D           2   
3      20000.0   36 months    0.0849       631.26     B          10   
4      10000.0   36 months    0.0649       306.45     A           6   

  home_ownership  annual_inc verification_status     issue_d  loan_status  \
0           RENT    110000.0        Not Verified  2015-12-01   Fully Paid   
1           RENT     70000.0        Not Verified  2015-12-01   Fully Paid   
2       MORTGAGE     55000.0        Not Verified  2015-12-01  Charged Off   
3       MORTGAGE     85000.0        Not Verified  2015-12-01   Fully Paid   
4           RENT     85000.0        Not Verified  2015-12-01   Fully Paid   

          purpose addr_state    dti  delinq_2yrs earliest_cr_line  \
0     credit_card         IL  12.45          0.0       2007-06-01   
1           house         FL  22.21          0.0       2004-07-01   
2     credit_card         CO  35.70          0.0       2001-04-01   
3  major_purchase         SC  17.61          1.0       1999-02-01   
4     credit_card         PA  13.07          0.0       2002-04-01   

   fico_range_low  inq_last_6mths  mths_since_last_delinq  open_acc  pub_rec  \
0           690.0             0.0                     NaN       8.0      0.0   
1           680.0             0.0                    38.0      24.0      0.0   
2           685.0             0.0                     NaN      14.0      0.0   
3           705.0             0.0                     3.0       8.0      0.0   
4           685.0             1.0                     NaN      14.0      1.0   

   pub_rec_bankruptcies  revol_bal  revol_util  total_acc  total_pymnt  \
0                   0.0    21374.0       0.845       12.0  24889.01336   
1                   0.0    19077.0       0.366       63.0  21780.58678   
2                   0.0    38623.0       0.780       28.0   5558.20000   
3                   0.0      826.0       0.057       15.0  21538.50898   
4                   1.0    10464.0       0.345       23.0  10998.97157   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0         24889.01         20000.00        4889.01                 0.0   
1         21780.59         20000.00        1780.59                 0.0   
2          5558.20          2687.15        2871.05                 0.0   
3         21538.51         20000.00        1538.51                 0.0   
4         10998.97         10000.00         998.97                 0.0   

   collections_12_mths_ex_med application_type  acc_now_delinq  \
0                         0.0       Individual             0.0   
1                         0.0       Individual             0.0   
2                         0.0       Individual             0.0   
3                         0.0       Individual             0.0   
4                         0.0       Individual             0.0   

   chargeoff_within_12_mths  acc_open_past_24mths  avg_cur_bal  delinq_amnt  \
0                       0.0                   3.0       5356.0          0.0   
1                       0.0                   9.0       2759.0          0.0   
2                       0.0                   4.0      20578.0          0.0   
3                       0.0                   4.0      17700.0          0.0   
4                       0.0                   7.0       1997.0          0.0   

   mo_sin_old_rev_tl_op  mo_sin_rcnt_rev_tl_op  mo_sin_rcnt_tl  mort_acc  \
0                 102.0                   16.0            10.0       0.0   
1                 137.0                    2.0             2.0       1.0   
2                 176.0                    7.0             6.0       6.0   
3                  55.0                   32.0            13.0       3.0   
4                 129.0                    1.0             1.0       1.0   

In [9]:
# fix loan status column 

status_count = pd.DataFrame(lending['loan_status'].value_counts())

# reset and fix index and column labels 
status_df = status_count.reset_index()

status_df.rename(columns={'loan_status':'count'}, inplace = True)
status_df.rename(columns={'index':'loan_status'}, inplace = True)

# next add % of total 

total_loans = len(lending)

status_df['per_of_total'] = status_df['count'] / total_loans 

status_df




loan_status   count  per_of_total
0          Fully Paid  297817      0.786802
1         Charged Off   70633      0.186605
2             Current    9571      0.025286
3     In Grace Period     253      0.000668
4  Late (31-120 days)     156      0.000412
5   Late (16-30 days)      76      0.000201
6             Default      10      0.000026

* ~97.5% fall into either fully paid or charged off. 
* Given that we're just trying to identify customers that will pay off their defaults, the other status aren't relevant 
* In grace period, late and at official default could be useful as far as potentially predicting early warnings of issues, but the number of loans in that category is far too small to be useful 

In [10]:
# isolate just statuses of fully paid or charged off. 

statuses = ['Fully Paid', 'Charged Off']
lending.loan_status.isin(statuses)
lending = lending[lending.loan_status.isin(statuses)]

lending.shape



(368450, 69)

In [11]:
# let's assess the charge off rate with just loans that been deeemed uncollectible
# "charge off" or were fully paid 

status_count = pd.DataFrame(lending['loan_status'].value_counts())

# reset and fix index and column labels 
status_df = status_count.reset_index()

status_df.rename(columns={'loan_status':'count'}, inplace = True)
status_df.rename(columns={'index':'loan_status'}, inplace = True)

# next add % of total 

total_loans = len(lending)

status_df['per_of_total'] = status_df['count'] / total_loans 

status_df






loan_status   count  per_of_total
0   Fully Paid  297817      0.808297
1  Charged Off   70633      0.191703

A rather high 19.1% of the loans were defaulted on, which makes this a risky proposition for investors.

If you were to invest 10k in a bucket of loans that roughly approximated the performance of all the loans in this dataset at the average interest rate, a lot of your gains would be offset by principle losses. 

A more detailed analysis on potential gains and losses will follow once the data is fully cleaned and more high level EDA completed. 



In [12]:
# need to assess how to handle the NaNs in months since last delinquency. 


test = lending.dropna(subset=['mths_since_last_delinq'])

test.shape




(190503, 69)

In [13]:
# let's look at the default rates in the group without NaNs 


status_count_test = pd.DataFrame(test['loan_status'].value_counts())

# reset and fix index and column labels 
status_df_test = status_count_test.reset_index()

status_df_test.rename(columns={'loan_status':'count'}, inplace = True)
status_df_test.rename(columns={'index':'loan_status'}, inplace = True)

# next add % of total 

total_loans_test = len(test)

status_df_test['per_of_total'] = status_df_test['count'] / total_loans_test 

status_df_test




loan_status   count  per_of_total
0   Fully Paid  152658      0.801342
1  Charged Off   37845      0.198658

Given that nearly 1/2 of our dataset has NaN values for months since last delinquency, we have to study this column further as dropping all of these rows would probably hurt our analysis, as would filling in the values with the average. We're going to study this more to make a higher quality inference around what these values should be. 

In [14]:
# let's check and see if there are any zero values 

zero_delinq = lending[(lending['mths_since_last_delinq'] == 0)]
zero_delinq.shape



(434, 69)

In [15]:
# let's keep the rows with NaN values for months since last Delinquency 
# and see how they perform as far as paying off thier loans 

# let's look at the default rates in the group without NaNs 

count_null = lending[lending['mths_since_last_delinq'].isnull()]


status_count_null = pd.DataFrame(count_null['loan_status'].value_counts())

# reset and fix index and column labels 
status_df_null = status_count_null.reset_index()


status_df_null.rename(columns={'loan_status':'count'}, inplace = True)
status_df_null.rename(columns={'index':'loan_status'}, inplace = True)

# next add % of total 

total_loans_null = len(count_null)

status_df_null['per_of_total'] = status_df_null['count'] / total_loans_null

status_df_null




loan_status   count  per_of_total
0   Fully Paid  145159      0.815743
1  Charged Off   32788      0.184257

In [16]:
# since the charge off rates are rather close, let's see if there is a difference in average credit score


count_null['fico_range_low'].mean()



703.1405699449837

In [17]:
test['fico_range_low'].mean()

684.8219188149268

In [18]:
lending['fico_range_low'].mean()

693.6691138553399

The FICO scores are:

* 703 for the customers with NaN for months since last delinquency 
* 684 for the customers without a NaN, nearly all of whom have a value for time since last delinquency
* 693 for the customer group overall. 

The NaN group having a higher FICO does support the idea that since there wasn't a delinquency on their credit report, LC's systems simply didn't enter in a value and thus a NaN was generated. 

In [19]:
# let's look at interest rates for both groups as well 

count_null['int_rate'].mean()



0.12284871338095049

In [20]:
test['int_rate'].mean()

0.12758449000803132

In [21]:
# see if the group with NaN have lower instances of having ever had a significantly past due accounts 

count_null['num_accts_ever_120_pd'].sum()


2593.0

In [22]:
test['num_accts_ever_120_pd'].sum()

191197.0

**Findings Summary**

* Nearly 1/2 of the total dataset has "NaN" for months since last delinquency 
* Default rates for both the NaN and non NaN cohorts are nearly the same 19.18% and 18.46% respectively
* The credit scores for the group with NaN are nearly an entire tier higher, as they're nearly 20 points higher than the group without NaNs. 
* interest rates for the group without NaNs is about 1/2 a point higher.
* The group without NaNs have had 201,053 instances (for 199k loans) of having an account seriously past due (120 days) vs. only 2,690 for the NaN group of roughly ~184k loans. 

Given the evidence, it's reasonable to presume that the group with NaNs have significantly better credit and payment history and that the NaNs represent not having had a delinquency. The solution for this will be as follows:

* Remove the zeros as such a small number (457) that it won't hurt our analysis 
* fill in all the missing values with zeros 
* subtract 84 from all the other values and store that result in a column called "countdown_zero_delinq", as that way a value of 1 month since last delinquency isn't seen by the model as only slightly better than 1 and a lot worse than 75, when in fact 75 is a lot better than one and 0 is best of all  

Note: using 84 as delinquencies only stay on one's credit report for 84 months

In [23]:

# filter out the zero values 

lending = lending[(lending['mths_since_last_delinq'] != 0)]


# fill in the NaNs with zeroes 

lending.loc[:,'mths_since_last_delinq'].fillna(0, inplace=True)


# create the column with the countdown to zero delinquencies 

lending.loc[:, 'countdown_zero_delinq'] =\
lending.loc[:,'mths_since_last_delinq'].apply(lambda x: 0 if x == 0 else x - 84)


In [24]:
# see how many zeroes are in the months since recent revolving delinquency column 

zero_revol_delinq = lending[(lending['mths_since_recent_revol_delinq'] == 0)]
zero_revol_delinq.shape



(2, 70)

In [25]:
# repeat the process above for the column 'mths_since_recent_revol_delinq' as the same idea applies, if
# not more since there are even fewer zeroes (2 vs. 457)

# filter out the zero values 

lending = lending[(lending['mths_since_recent_revol_delinq'] != 0)]


# fill in the NaNs with zeroes 

lending.loc[:,'mths_since_recent_revol_delinq'].fillna(0, inplace=True)


# create the column with the countdown to zero delinquencies 

lending.loc[:, 'countdown_zero_revol_delinq'] =\
lending.loc[:,'mths_since_recent_revol_delinq'].apply(lambda x: 0 if x == 0 else x - 84)




In [26]:
lending.shape

(368014, 71)

In [27]:
# evaluate the rest of the columns for potential NaNs 


print(" \nCount total NaN at each column in a DataFrame : \n\n", 
      lending.isnull().sum()) 

 
Count total NaN at each column in a DataFrame : 

 funded_amnt                          0
term                                 0
int_rate                             0
installment                          0
grade                                0
emp_length                           0
home_ownership                       0
annual_inc                           0
verification_status                  0
issue_d                              0
loan_status                          0
purpose                              0
addr_state                           0
dti                                  0
delinq_2yrs                          0
earliest_cr_line                     0
fico_range_low                       0
inq_last_6mths                       0
mths_since_last_delinq               0
open_acc                             0
pub_rec                              0
pub_rec_bankruptcies                 0
revol_bal                            0
revol_util                           0
total_acc  

In the original pass the following columns all had the exact same number of NaNs 367,120:
* all_util
* inq_last_12m
* max_bal_bc
* open_acc_6m
* open_il_12m
* open_act_il 

Given the consistency of the number of NaNs it stands to reason that this data field wasn't being tracked at the time 90% of this data was generated. Since it's not present for 367k out of 384k records there is no value in adding zero, so I updated the pre-processing file not to select these columns.  


In [28]:
# see the shape of the final data frame 

lending.shape

(368014, 71)

In [29]:
# write the final data frame to csv file for use in the machine learning and additional EDA notebooks

lending.to_csv('data/LC_2015_clean(4).csv', date_format='%Y-%m-%d', index=False)

